### ISSUE TITLE
buf_write does not properly handle non-US-ASCII filename extension

### ISSUE THREAD

**Issue by Safari77** [vim/vim#14948]
### Steps to reproduce

1) edit filename with UTF8 chars in the extension, like x.txt.😦
2) :w
3) set backup
4) :w 
5) set nobackup
6) :w
7) when writing and filesystem does not allow invalid UTF8:
```
[pid 3328321] 16:36:30.038496 newfstatat(AT_FDCWD, "x.txt.\360\237\230\246~", {st_dev=makedev(0, 0x1f), st_ino=26423526, st_mode=S_IFREG|0664, st_nlink=1, st_uid=500, st_gid=500, st_blksize=512, st_blocks=2, st_size=8, st_atime=1717939887 /* 2024-06-09T16:31:27.126006617+0300 */, st_atime_nsec=126006617, st_mtime=1717939887 /* 2024-06-09T16:31:27.126006617+0300 */, st_mtime_nsec=126006617, st_ctime=1717939887 /* 2024-06-09T16:31:27.131006744+0300 */, st_ctime_nsec=131006744}, 0) = 0 <0.000018>
[pid 3328321] 16:36:30.038603 newfstatat(AT_FDCWD, "x.txt.\360\237\230z~", 0x7ffe15901c30, 0) = -1 EILSEQ (Invalid or incomplete multibyte or wide character) <0.000028>
```
Note: vim manages to write the file eventually, but the filename handling should be fixed anyways.

### Expected behaviour

as buf_write() is 2040 lines, I get headache from reading it, but it writes 'z' into a char pointer. A different method maybe could be used which is compatible with UTF8 (or any charset).

### Version of Vim

9.1.0470

### Environment

Fedora 39
kitty 0.35.1

**Comment by chrisbra** [comment]
> Note: vim manages to write the file eventually, but the filename handling should be fixed anyways.

So is the file extension written correctly or not?

**Comment by Safari77** [comment]
> > Note: vim manages to write the file eventually, but the filename handling should be fixed anyways.
> 
> So is the file extension written correctly or not?

No, it first does `rename("x.txt.\360\237\230\246", "x.txt.\360\237\230z~")` which fails, then tries to 
`rename("x.txt.\360\237\230\246", "/home/safari/tmp/x.txt.\360\237\230\246~")` which also fails, then 
deletes the original file  `unlink("x.txt.\360\237\230\246") = 0` which is a big no no, then finally writes the file in a non-atomic way in place.

(I am working on a patch for vim which writes files atomically.)
